feat: store users languages and skills problem count data

svn27 opened the pull request:
## Changes
- Created new data classes for skill tags and languages.
- Edited the query in `fetch_problems_solved_and_rank` to query the required data.
- Added new BaseModels to `user.py` for the new data.
- Added new attributes to Record `record.py` for the new BaseModels.
- Edited `update_stats` to provide the new data into the Record constructor.
- Imported the new BaseModels in `database/models/__init__.py`

Kevin-Roman commented on utils/problems.py:
```diff
@@ -415,6 +451,34 @@ async def fetch_problems_solved_and_rank(
+        intermediate_tags_count = tags_problem_count["intermediate"]
+        fundamental_tags_count = tags_problem_count["fundamental"]
+
+        languages_problem_count_list = []
+
+        for language_problem_count in languages_problem_count:
+            languages_problem_count_list.append(LanguageProblemCount(language = language_problem_count["languageName"], problem_count = language_problem_count["problemsSolved"]))
```
This seems like an appropriate place to use list comprehensions, similarly for the advanced_tag_list , etc...

Kevin-Roman commented on utils/problems.py:
```diff
@@ -415,6 +451,34 @@ async def fetch_problems_solved_and_rank(
         submit_stats_global = matched_user["submitStatsGlobal"]
         ac_submission_num = submit_stats_global["acSubmissionNum"]
 
+        languages_problem_count = matched_user["languageProblemCount"]
+
+        tags_problem_count = matched_user["tagProblemCounts"]
+        advanced_tags_count = tags_problem_count["advanced"]
+        intermediate_tags_count = tags_problem_count["intermediate"]
+        fundamental_tags_count = tags_problem_count["fundamental"]
```
Defining each of these variables is unecessary. Just remove these, and just refer to the actually value (aka tags_problem_count["fundamental"], wherever necessary)

Kevin-Roman commented on utils/problems.py:
```diff
@@ -415,6 +451,34 @@ async def fetch_problems_solved_and_rank(
+        intermediate_tags_count = tags_problem_count["intermediate"]
+        fundamental_tags_count = tags_problem_count["fundamental"]
+
+        languages_problem_count_list = []
```
There shouldn't really be a need to add the type name to the variable name (in this case `list`). Also, to maintain consistency with the naming of the graphql query, just call this variable `language_problem_count`. Similarly for the rest, e.g. your `advanced_tag_list` variable, rename to `tagProblemCountsAdvanced`

Kevin-Roman commented on utils/problems.py:
```diff
@@ -415,6 +451,34 @@ async def fetch_problems_solved_and_rank(
+        advanced_tag_list = []
+
+        for advanced_tag_count in advanced_tags_count:
+            advanced_tag_list.append(SkillProblemCount(skill = advanced_tag_count["tagName"], problem_count = advanced_tag_count["problemsSolved"]))
```
As I said through discord, install and use a python formatter

Kevin-Roman commented on utils/problems.py:
```diff
@@ -47,12 +47,25 @@ class Submissions:
     real_name: str
     submissions: Submissions
-
+    languages_problem_count: list #Made up of LanguageProblemCount objects
+    fundamental_skills_problem_count: list #Made up of SkillProblemCount objects
+    intermediate_skills_problem_count: list #Made up of SkillProblemCount objects
+    advanced_skills_problem_count: list #Made up of SkillProblemCount objects
```
rename the to match my suggestions from below, e.g. `tag_problem_counts_advanced`. We could start the variable name with the difference (aka the difficulty such as advanced, intermediate, fund...), but as per my suggestion, I think in this case we should start the class attribute names with the common part of the name. I hope that makes sense, pretty waffly explanation from me so mb

Kevin-Roman replied:
No need to add those comments. We can write self-documenting code, by actually type hint what the list is made up of, which is cool! So rather than type hint just `: list`, say `: list[LanguageProblemCount]`. Just for your knowledge, type hints aren't actually ever utilised by python during compilation, they're merely there to help programmers as they provide readable insights to the variable's type, but also aid in your IDE's intellisense.

svn27 commented:
- Changed variable names to better reflect the graphql queries in `problems.py`
- Added new attributes to Record `record.py`
- Added new BaseModels to `user.py` 
- Edited `update_stats ` to store the new data
- Added new imports to `__init__.py`

Kevin-Roman reviewed (approved):
LGTM!
